internal/pkg/image/driver/gocryptfsdriver.go

```diff
@@ -0,0 +1,464 @@
+	if err != nil {
+		return err
+	}
+	g.decryptInfo = &decryptInfo{
+		cipherDir: filepath.Join(tmpDir, "cipher"),
+		plainDir:  filepath.Join(tmpDir, "plain"),
+		pass:      "",
+		confPath:  filepath.Join(filepath.Join(tmpDir, "cipher"), "gocryptfs.conf"),
+	}
+	g.cleanFn = func() error {
+		return os.RemoveAll(tmpDir)
+	}
+
+	err = os.Mkdir(g.decryptInfo.cipherDir, 0o700)
+	if err != nil {
+		return err
+	}
+	err = os.Mkdir(g.decryptInfo.plainDir, 0o700)
+	if err != nil {
+		return err
+	}
+
+	g.mountInfos = make([]mountInfo, 0)
+	return nil
+}
+
+func (g *gocryptfsFeature) Mount(params *image.MountParams, mfunc image.MountFunc, cmdPrefix []string) error {
+	// step 1. initialize the gocryptfs
+	if g.gocryptfsPath == "" {
+		return fmt.Errorf("gocryptfs is required")
+	}
+	if g.squashfusePath == "" {
+		return fmt.Errorf("squashfuse | squashfuse_ll is required")
+	}
+
+	err := g.create()
+	if err != nil {
+		return err
+	}
+
+	g.decryptInfo.pass = string(params.Key)
+
+	ch := make(chan error, 1)
+
+	// step 2. directly mount the sif with data partition offset to cipher folder
+
+	oldDest := params.Target
+	var extraFiles []*os.File
+	if path.Dir(params.Source) == "/proc/self/fd" {
+		extraFiles = make([]*os.File, 1)
+		targetFd, _ := strconv.Atoi(path.Base(params.Source))
+		extraFiles[0] = os.NewFile(uintptr(targetFd), params.Source)
+	}
+	cmd, err := g.cmdStart(cmdPrefix, sysAttr, nil, extraFiles, g.squashfusePath, "-f", "-o", fmt.Sprintf("uid=%v,gid=%v,offset=%v", os.Getuid(), os.Getgid(), params.Offset), params.Source, g.decryptInfo.cipherDir)
+	if err != nil {
+		return err
+	}
+
+	go CheckMountInfo(g.decryptInfo.cipherDir, ch)
+
+	err = <-ch
+	if err != nil {
+		return err
+	}
+
+	// verify mounted files
+	files, err := os.ReadDir(g.decryptInfo.cipherDir)
+	if err != nil || len(files) == 0 {
+		return fmt.Errorf("previous squashfuse mount failed, because there are no files in %s", g.decryptInfo.cipherDir)
+	}
+
+	var targetfile string
+	for _, file := range files {
+		info, err := file.Info()
+		if err != nil {
+			return err
+		}
+		sylog.Debugf("file name: %s, file size: %d\n", info.Name(), info.Size())
+		if info.Size() == 0 {
+			return fmt.Errorf("%s file size should not be 0", file.Name())
+		}
+		if strings.HasPrefix(file.Name(), "squashfs-") {
+			targetfile = file.Name()
+			break
+		}
+	}
+
+	if targetfile == "" {
+		return fmt.Errorf("could not locate the squashfuse created file, previous step might fail")
+	}
+
+	g.mountInfos = append(g.mountInfos, mountInfo{
+		pid:       cmd.Process.Pid,
+		srcPath:   params.Source,
+		dstPath:   g.decryptInfo.cipherDir,
+		mountType: SQUASHFS,
+	})
+
+	// step 3. trigger gocryptfs fuse mount
+	cmd, err = g.cmdStart(cmdPrefix, sysAttr, strings.NewReader(fmt.Sprintf("%s\n", g.pass)), nil, g.gocryptfsPath, g.decryptInfo.cipherDir, g.decryptInfo.plainDir)
+	if err != nil {
+		return err
+	}
+
+	go CheckMountInfo(g.decryptInfo.plainDir, ch)
+	err = <-ch
+	if err != nil {
+		return err
+	}
```

Comment:
I don't understand the point of using "go" here if it immediately waits for the response, and likewise below.